### Table of Contents
01. Import Libraries
02. Import Data
03. Consistency Checks for df_prods
04. Consistency Checks for df_ords
05. Export Dataframes

# 01. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
# Folder Path
path = r'C:\Users\jrper\OneDrive\Documents\Career Foundry Data Analytics Program\Instacart Basket Analysis'

In [3]:
# Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Consistency Checks for df_prods

## Explore Data

In [5]:
# View df_prods first 5 rows
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [6]:
# View df_prods last 5 rows
df_prods.tail()

product_id                                product_name  aisle_id  \
49688       49684   Vodka, Triple Distilled, Twist of Vanilla       124   
49689       49685          En Croute Roast Hazelnut Cranberry        42   
49690       49686                            Artisan Baguette       112   
49691       49687  Smartblend Healthy Metabolism Dry Cat Food        41   
49692       49688                      Fresh Foaming Cleanser        73   

       department_id  prices  
49688              5     5.3  
49689              1     3.1  
49690              3     7.8  
49691              8     4.7  
49692             11    13.5

In [7]:
# Column names for df_prods
df_prods.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

In [8]:
# Check df_prods size
df_prods.shape

(49693, 5)

In [9]:
# Check data types for df_prods
df_prods.dtypes

product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

## Check for mixed-type data

In [10]:
# Check for mixed-type columns in df_prods
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


#### Observations: 'product_name' column has mixed-type data. Will address after checking for missing values.

## Check for missing values

In [11]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

#### Observations: There are 16 missing values in the 'product_name' column

In [12]:
# Create subset containing missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
# View missing values subset
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

#### Observations: Records with missing product names will be filtered out and a new dataframe created.

In [14]:
# Find row and column numbers for df_prods
df_prods.shape

(49693, 5)

In [15]:
# Create dataframe without missing values in 'product_name' column
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
# Compare row and column numbers for new dataframe
df_prods_clean.shape

(49677, 5)

#### Observations: df_prods_clean has 16 fewer rows than df_prods, so missing values have been successfully filtered out.

In [17]:
# Check again for mixed-type data after filtering out missing values
for col in df_prods_clean.columns.tolist():
  weird = (df_prods_clean[[col]].applymap(type) != df_prods_clean[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods_clean[weird]) > 0:
    print (col)

#### Observations: No mixed-type data columns detected after filtering out missing values.

## Check for duplicates

In [18]:
# Find full duplicates for df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
# View duplicates subset
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

#### Observations: 5 full duplicates detected and will be dropped

In [20]:
# Create dataframe with duplicates removed
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
# Compare row and column numbers for dataframe with duplicates removed
df_prods_clean_no_dups.shape

(49672, 5)

#### Observations: 5 fewer records means that duplicates have been successfully dropped.

## Descriptive Statistics

In [22]:
# Descriptive Statistics for df_prods_clean_no_dups
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### Observations: Max value for 'prices' seems high and should be investigated.

In [23]:
# Investigate record with max 'prices' value
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

#### Observations: This price is an error or a placeholder value and must be resolved before further analysis.

In [24]:
# Check for products that cost more than $50
df_prods_clean_no_dups[df_prods_clean_no_dups["prices"]> 50.0]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

#### Observations: These prices are obviously incorrect. The price for Lowfat 2% Milkfat Cottage Cheese will be replaced with 1.49000, and the price for 2% Reduced Fat Milk will be replaced with 9.99990.

In [25]:
# Replace 'prices' value for 'product_id' 21553 with 1.49000 and 'product_id' 33664 with 9.99  
df_prods_clean_no_dups = df_prods_clean_no_dups.replace ({"prices":{99999.0: 9.99, 14900.0:1.49 }})

In [26]:
# Check that changes have been successfully made
df_prods_clean_no_dups[df_prods_clean_no_dups["product_id"]==21553]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   

       prices  
21554    1.49

In [27]:
# Check that changes have been successfully made
df_prods_clean_no_dups[df_prods_clean_no_dups["product_id"]==33664]

product_id           product_name  aisle_id  department_id  prices
33666       33664  2 % Reduced Fat  Milk        84             16    9.99

#### Observations: Changes have been successfully made.

# 04. Consistency Checks for df_ords

## Explore Data

In [28]:
# Drop unnamed column for df_ords
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [29]:
# View df_ords first 5 rows
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [30]:
# View df_ords last 5 rows
df_ords.tail()

order_id  user_id  order_number  orders_day_of_week  \
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0

In [31]:
# Column names for df_ords
df_ords.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [32]:
# Check df_ords size
df_ords.shape

(3421083, 6)

In [33]:
# Check data types for df_ords
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

## Check for mixed-type data

In [34]:
# Step 3: Check for mixed-type data in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### Observations: No mixed-type data detected for df_ords

## Check for missing values

In [35]:
# Step 5: Check for Missing Values in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [36]:
# Create subset to show missing values in df_ords
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [37]:
# View missing values subset
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

#### Observations: All of the records containing missing values in the 'days_since_prior_order' column also have an 'order_number' value of 1. Because the customer is ordering for the first time, there won't be a value showing how long it has been since a prior order. The maximum value in the 'user_id' column is 206209, which coincides with the number of missing values in the 'days_since_prior_order' column: one for each new customer.

In [38]:
# Step 6: Address missing values in 'days_since_prior_order' column

#### I have chosen not to make any changes to the missing values in the 'days_since_prior_order' column. It doesn't make sense to impute values or remove/filter the data. Statistics can still be calculated with the NaN values present. A value of 1 in the 'order_number' column can act as a flag for a new customer.

## Check for duplicates

In [39]:
# Step 7: Check for duplicate values in df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [40]:
# View df_ords_dups 
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### Observations: No duplicates detected.

## Descriptive Statistics

In [41]:
# Step 2: Descriptive Statistics for df_ords
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### Observations: Descriptive statistics for df_ords appear to be normal.

# 05. Export Dataframes

In [42]:
# Export df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [45]:
# Step 9: Export final df_ords 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index = False)